Fix ambiguity of atomic accessor operator[].

fodinabor opened the pull request:
As reported in https://github.com/illuhad/hipSYCL/issues/380#issuecomment-803572968 [1], the `operator[]` is ambiguous, if using an atomic accessor.

fodinabor committed:
Fix ambiguity of atomic accessor operator[].

illuhad commented:
Actually we implement the interface exactly as described by the SYCL 2020 spec :)
https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_interface_for_buffer_command_accessors

It seems that there is now an oversight in the spec w.r.t. to the conditions in which to enable `operator[]`, presumably because `atomic` is completely deprecated...

illuhad commented on tests/sycl/accessor.cpp:
```diff
@@ -161,6 +161,17 @@ BOOST_AUTO_TEST_CASE(accessor_api) {
+    cgh.single_task<class accessor_api_atomic_device_accessors>([atomicAcc](){
+      atomicAcc[0].exchange(0);
+    });
+  });
```
Philosophical question: Do we actually need to/Should we test deprecated features?

fodinabor replied:
I think as long as we did not remove them, we should test it, as we still provide it and the user will expect it to work as long as it's there.

illuhad commented on include/hipSYCL/sycl/libkernel/accessor.hpp:
```diff
@@ -537,15 +537,17 @@ class accessor : public detail::accessor_base<std::remove_const_t<dataT>> {
   template<int D = dimensions,
-            std::enable_if_t<(D == 1), bool> = true>
+            access::mode M = accessmode,
+            std::enable_if_t<(D == 1) && (M != access::mode::atomic), bool> = true>
```
Wouldn't local accessor be affected as well?

fodinabor replied:
only if they had atomic support in the first case (which they [deprecatedly should have as of SYCL 2020](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_accessor_specialization_with_targetlocal))

fodinabor reviewed (commented):
The spec seems to be undecided how to deal with this..
```cpp
/// note the missing accessMode check
  /* Available only when: (dimensions > 0) */
  reference operator[](id<dimensions> index) const;

/// note the missing accessMode check
  /* Available only when: (dimensions > 1) */
  __unspecified__ &operator[](size_t index) const;

/// note the accessMode check
  /* Available only when: (accessMode != access_mode::atomic && dimensions == 1) */
  reference operator[](size_t index) const;

  /* Deprecated
  Available only when: (accessMode == access_mode::atomic && dimensions ==  0) */
  operator cl::sycl::atomic<dataT, access::address_space::global_space> () const;

  /* Deprecated
  Available only when: (accessMode == access_mode::atomic && dimensions == 1) */
  cl::sycl::atomic<dataT, access::address_space::global_space> operator[](
    id<dimensions> index) const;
```

Just had another look.. seems we're missing atomic support for multi-dimensional accessors in `subscript_proxy` in general..

Thus.. do we plan on keeping atomics around? If so, I will add the missing parts for local and subscript proxy

fodinabor committed:
Add deprecation note for atomic accessors and fix multi dimensional atomic accessors and local atomic accessors.

fodinabor commented:
@hipsycl-ci-bot test this now as well, please.

fodinabor committed:
Use nd_range parallel_for for accessor_api_atomic_device_accessors test.

Local memory doesn't make too much sense in single_task and is also not allowed per SYCL spec.

illuhad commented:
@fodinabor Yes, I think when rewriting the spec we didn't really think about if/how things should play with deprecated stuff in this particular case...

I can get behind continuing to support for now what we already have, but not sure if it's a good idea to complicate `subscript_proxy` just for some deprecated feature...

EDIT: Okay, I see you already did the `subscript_proxy`. Nevermind then, so be it ;)

illuhad reviewed (approved):
And I was so happy that I could remove all those checks when I changed the interfaces to SYCL 2020 definitions... Can't wait for atomic accessors to go away for good :)
